### ISSUE TITLE
Kernel Hook stops working after 20 minutes 

### ISSUE THREAD

**Issue by chico-depressive** [EasyHook/EasyHook#161]
@spazzarama I am getting a very weird behaviour, myFuncHook gets called, but after about 20 minutes it is called only sometimes, sometimes it's not, take a look at the code:

```
       ULONG ACLEntries[1] = { 0 };
	NTSTATUS result = STATUS_INTERNAL_ERROR;
	
	result = LhInstallHook(
		target_func_prologue,
		fake_func_prologue,
		(PVOID)0x12345678,
		pHHook);
	LhSetGlobalExclusiveACL(ACLEntries, 1);

        return result;
```

I also changed easyHook source code, forcing IsProcessIntercepted to "return TRUE" everytime
Do you have any ideas why I am getting this weird behaviour?

**Comment by chico-depressive** [comment]
I am making some runtime analisys and found this inside LhBarrierIntro:

```
// open pointer table
	Exists = TlsGetCurrentValue(&Unit.TLS, &Info);

	if(!Exists)
	{
		if (!TlsAddCurrentThread(&Unit.TLS)) {
			DbgPrint("!TlsAddCurrentThread \n");
			return FALSE;
		}
	}
```

after 20 minutes this line gets called -> DbgPrint("!TlsAddCurrentThread \n");
any idea what is going on?

**Comment by spazzarama** [comment]
@chico-depressive If TlsAddCurrentThread returns false, then I'm pretty sure it means that the MAX_THREAD_COUNT has been exceeded, which is curently set to 128.

I haven't had a good look at it, but perhaps try increasing this and see if that changes what you are experiencing.

**Comment by chico-depressive** [comment]
@spazzarama thank you for your help,

There is any problem if I remove TlsAddCurrentThread and TlsGetCurrentValue lines?

It looks like threads are being added, but not removed, can I raise this value to something like 12800?

**Comment by spazzarama** [comment]
@chico-depressive does this happen with any hook? I think that just increasing the limit will only delay the issue, instead we should see if we can determine the root cause of the issue.

**Comment by chico-depressive** [comment]
@spazzarama given that my kernel hook is monitoring ALL threads in the system, is really 128 a good number? 

If 128 processes(or threads) are created and call myHookFunc then probably TlsAddCurrentThread will fail

I am going to try to raise it to 1024 and I will post the results

**Comment by spazzarama** [comment]
@chico-depressive yeah I think it will need to be increased. However you mentioned that they are never being released? Is that the case? 

Hmm I guess for a kernel hook it isn't going to be attached/detached in the same manner (I think the current implementation is relying upon thread detached in DLLMain).

**Comment by chico-depressive** [comment]
@spazzarama tested and works, but as I have two problems:

1  - if a lot of threads starts calling the hooked function it starts failing
2 - windows x64 does not allows unsigned drivers to set some callbacks (thread creation and process creation)

to solve that problems, please consider creating some global hooking mode thread independent, something like:
```
hook.SetGlobal(true);
```
thank you

**Comment by chico-depressive** [comment]
@spazzarama can you help me to implement "hook.SetGlobal(true);" feature? It looks like too complicated to set global hooks thread independent I don't think I have that skill

**Comment by spazzarama** [comment]
@chico-depressive you mean within driver project right? The issue you are hitting is that EasyHook has a deadlock barrier to try and prevent deadlocks caused by threading and your hook.

I'm not sure when I will get a chance to look at this. Perhaps to help could you provide a test scenario so that the problem can be easily reproduced?
